### ISSUE TITLE
Build errors from platformio

### ISSUE THREAD

**Issue by mr-sneezy** [gztproject/NixieClock#1]
I've had some PCB's made by JLCPCB and am now trying to build the code for a D1 Mini.

Got a lot of PlatformIO build errors, more that my very limited Arduino experience can cope with. Looks like I'm doing something wrong, anyone got an idea on how to clear some of these?
The first is this error: 'localtime' was not declared in this scope in the NTP.cpp

Processing d1_mini (platform: espressif8266; board: d1_mini; framework: arduino)
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

Verbose mode can be enabled via `-v, --verbose` option
CONFIGURATION: https://docs.platformio.org/page/boards/espressif8266/d1_mini.html
PLATFORM: Espressif 8266 (1.6.0) > WeMos D1 R2 & mini
HARDWARE: ESP8266 80MHz, 80KB RAM, 4MB Flash
PACKAGES:
 - framework-arduinoespressif8266 1.20400.1 (2.4.0)  
 - tool-esptool 1.409.0 (4.9)
 - toolchain-xtensa 1.40802.0 (4.8.2)
LDF: Library Dependency Finder -> http://bit.ly/configure-pio-ldf
LDF Modes: Finder ~ chain, Compatibility ~ soft
Found 31 compatible libraries
Scanning dependencies...
Dependency Graph
|-- <ShiftRegister74HC595> 1.3.1
|-- <MyWiFi>
|   |-- <ESP8266WiFi> 1.0
|-- <NixieClock>
|   |-- <ShiftRegister74HC595> 1.3.1
|-- <NTP>
|   |-- <ESP8266WiFi> 1.0
|-- <OTA>
|   |-- <ArduinoOTA> 1.0
|   |   |-- <ESP8266WiFi> 1.0
|   |   |-- <ESP8266mDNS>
|   |   |   |-- <ESP8266WiFi> 1.0
Building in release mode
Compiling .pio\build\d1_mini\lib9c0\NTP\NTP.cpp.o
Compiling .pio\build\d1_mini\FrameworkArduino\Print.cpp.o
Compiling .pio\build\d1_mini\FrameworkArduino\Schedule.cpp.o
Compiling .pio\build\d1_mini\FrameworkArduino\Stream.cpp.o
Compiling .pio\build\d1_mini\FrameworkArduino\StreamString.cpp.o
lib\NTP\src\NTP.cpp: In member function 'void NTP::CheckTime()':
lib\NTP\src\NTP.cpp:88:37: error: 'localtime' was not declared in this scope
     timeinfo = localtime (&localTime);    
                                     ^
lib\NTP\src\NTP.cpp: In member function 'int NTP::getSeconds()':
lib\NTP\src\NTP.cpp:151:20: error: invalid use of incomplete type 'struct tm'
     return timeinfo->tm_sec;
                    ^
In file included from c:\users\martin\.platformio\packages\toolchain-xtensa\xtensa-lx106-elf\include\c++\4.8.2\cwchar:44:0,
                 from c:\users\martin\.platformio\packages\toolchain-xtensa\xtensa-lx106-elf\include\c++\4.8.2\bits\postypes.h:40,
                 from c:\users\martin\.platformio\packages\toolchain-xtensa\xtensa-lx106-elf\include\c++\4.8.2\bits\char_traits.h:40,
                 from c:\users\martin\.platformio\packages\toolchain-xtensa\xtensa-lx106-elf\include\c++\4.8.2\string:40,
                 from c:\users\martin\.platformio\packages\toolchain-xtensa\xtensa-lx106-elf\include\c++\4.8.2\random:41,
                 from c:\users\martin\.platformio\packages\toolchain-xtensa\xtensa-lx106-elf\include\c++\4.8.2\bits\stl_algo.h:65,
                 from c:\users\martin\.platformio\packages\toolchain-xtensa\xtensa-lx106-elf\include\c++\4.8.2\algorithm:62,
                 from C:\Users\Martin\.platformio\packages\framework-arduinoespressif8266\cores\esp8266/Arduino.h:240,
                 from lib\NTP\src/NTP.h:6,
                 from lib\NTP\src\NTP.cpp:2:
C:\Users\Martin\.platformio\packages\framework-arduinoespressif8266\tools\sdk\libc\xtensa-lx106-elf\include/wchar.h:56:8: error: forward declaration of 'struct tm'
 struct tm;
        ^
lib\NTP\src\NTP.cpp: In member function 'int NTP::getMinutes()':
lib\NTP\src\NTP.cpp:155:20: error: invalid use of incomplete type 'struct tm'
     return timeinfo->tm_min;
                    ^
In file included from c:\users\martin\.platformio\packages\toolchain-xtensa\xtensa-lx106-elf\include\c++\4.8.2\cwchar:44:0,
                 from c:\users\martin\.platformio\packages\toolchain-xtensa\xtensa-lx106-elf\include\c++\4.8.2\bits\postypes.h:40,
Compiling .pio\build\d1_mini\FrameworkArduino\Tone.cpp.o
                 from c:\users\martin\.platformio\packages\toolchain-xtensa\xtensa-lx106-elf\include\c++\4.8.2\bits\char_traits.h:40,
Compiling .pio\build\d1_mini\FrameworkArduino\Updater.cpp.o
                 from c:\users\martin\.platformio\packages\toolchain-xtensa\xtensa-lx106-elf\include\c++\4.8.2\string:40,
                 from c:\users\martin\.platformio\packages\toolchain-xtensa\xtensa-lx106-elf\include\c++\4.8.2\random:41,
                 from c:\users\martin\.platformio\packages\toolchain-xtensa\xtensa-lx106-elf\include\c++\4.8.2\bits\stl_algo.h:65,
                 from c:\users\martin\.platformio\packages\toolchain-xtensa\xtensa-lx106-elf\include\c++\4.8.2\algorithm:62,
                 from C:\Users\Martin\.platformio\packages\framework-arduinoespressif8266\cores\esp8266/Arduino.h:240,
                 from lib\NTP\src/NTP.h:6,
                 from lib\NTP\src\NTP.cpp:2:
C:\Users\Martin\.platformio\packages\framework-arduinoespressif8266\tools\sdk\libc\xtensa-lx106-elf\include/wchar.h:56:8: error: forward declaration of 'struct tm'
 struct tm;
        ^
lib\NTP\src\NTP.cpp: In member function 'int NTP::getHours()':
lib\NTP\src\NTP.cpp:159:20: error: invalid use of incomplete type 'struct tm'
     return timeinfo->tm_hour;    
                    ^
In file included from c:\users\martin\.platformio\packages\toolchain-xtensa\xtensa-lx106-elf\include\c++\4.8.2\cwchar:44:0,
                 from c:\users\martin\.platformio\packages\toolchain-xtensa\xtensa-lx106-elf\include\c++\4.8.2\bits\postypes.h:40,
                 from c:\users\martin\.platformio\packages\toolchain-xtensa\xtensa-lx106-elf\include\c++\4.8.2\bits\char_traits.h:40,
                 from c:\users\martin\.platformio\packages\toolchain-xtensa\xtensa-lx106-elf\include\c++\4.8.2\string:40,
                 from c:\users\martin\.platformio\packages\toolchain-xtensa\xtensa-lx106-elf\include\c++\4.8.2\random:41,
                 from c:\users\martin\.platformio\packages\toolchain-xtensa\xtensa-lx106-elf\include\c++\4.8.2\bits\stl_algo.h:65,
                 from c:\users\martin\.platformio\packages\toolchain-xtensa\xtensa-lx106-elf\include\c++\4.8.2\algorithm:62,
                 from C:\Users\Martin\.platformio\packages\framework-arduinoespressif8266\cores\esp8266/Arduino.h:240,
                 from lib\NTP\src/NTP.h:6,
                 from lib\NTP\src\NTP.cpp:2:
C:\Users\Martin\.platformio\packages\framework-arduinoespressif8266\tools\sdk\libc\xtensa-lx106-elf\include/wchar.h:56:8: error: forward declaration of 'struct tm'
 struct tm;
        ^
lib\NTP\src\NTP.cpp: In member function 'int NTP::getDay()':
lib\NTP\src\NTP.cpp:163:20: error: invalid use of incomplete type 'struct tm'
     return timeinfo->tm_mday;
Compiling .pio\build\d1_mini\FrameworkArduino\WMath.cpp.o
                    ^
In file included from c:\users\martin\.platformio\packages\toolchain-xtensa\xtensa-lx106-elf\include\c++\4.8.2\cwchar:44:0,
                 from c:\users\martin\.platformio\packages\toolchain-xtensa\xtensa-lx106-elf\include\c++\4.8.2\bits\postypes.h:40,
                 from c:\users\martin\.platformio\packages\toolchain-xtensa\xtensa-lx106-elf\include\c++\4.8.2\bits\char_traits.h:40,
                 from c:\users\martin\.platformio\packages\toolchain-xtensa\xtensa-lx106-elf\include\c++\4.8.2\string:40,
                 from c:\users\martin\.platformio\packages\toolchain-xtensa\xtensa-lx106-elf\include\c++\4.8.2\random:41,
                 from c:\users\martin\.platformio\packages\toolchain-xtensa\xtensa-lx106-elf\include\c++\4.8.2\bits\stl_algo.h:65,
                 from c:\users\martin\.platformio\packages\toolchain-xtensa\xtensa-lx106-elf\include\c++\4.8.2\algorithm:62,
                 from C:\Users\Martin\.platformio\packages\framework-arduinoespressif8266\cores\esp8266/Arduino.h:240,
                 from lib\NTP\src/NTP.h:6,
                 from lib\NTP\src\NTP.cpp:2:
C:\Users\Martin\.platformio\packages\framework-arduinoespressif8266\tools\sdk\libc\xtensa-lx106-elf\include/wchar.h:56:8: error: forward declaration of 'struct tm'
 struct tm;
        ^
lib\NTP\src\NTP.cpp: In member function 'int NTP::getDow()':
lib\NTP\src\NTP.cpp:167:20: error: invalid use of incomplete type 'struct tm'
     return timeinfo->tm_wday + 1;
                    ^
In file included from c:\users\martin\.platformio\packages\toolchain-xtensa\xtensa-lx106-elf\include\c++\4.8.2\cwchar:44:0,
                 from c:\users\martin\.platformio\packages\toolchain-xtensa\xtensa-lx106-elf\include\c++\4.8.2\bits\postypes.h:40,
                 from c:\users\martin\.platformio\packages\toolchain-xtensa\xtensa-lx106-elf\include\c++\4.8.2\bits\char_traits.h:40,
                 from c:\users\martin\.platformio\packages\toolchain-xtensa\xtensa-lx106-elf\include\c++\4.8.2\string:40,
                 from c:\users\martin\.platformio\packages\toolchain-xtensa\xtensa-lx106-elf\include\c++\4.8.2\random:41,
                 from c:\users\martin\.platformio\packages\toolchain-xtensa\xtensa-lx106-elf\include\c++\4.8.2\bits\stl_algo.h:65,
                 from c:\users\martin\.platformio\packages\toolchain-xtensa\xtensa-lx106-elf\include\c++\4.8.2\algorithm:62,
                 from C:\Users\Martin\.platformio\packages\framework-arduinoespressif8266\cores\esp8266/Arduino.h:240,
                 from lib\NTP\src/NTP.h:6,
                 from lib\NTP\src\NTP.cpp:2:
C:\Users\Martin\.platformio\packages\framework-arduinoespressif8266\tools\sdk\libc\xtensa-lx106-elf\include/wchar.h:56:8: error: forward declaration of 'struct tm'
 struct tm;
        ^
lib\NTP\src\NTP.cpp: In member function 'int NTP::getMonth()':
lib\NTP\src\NTP.cpp:171:20: error: invalid use of incomplete type 'struct tm'
     return timeinfo->tm_mon + 1;
                    ^
In file included from c:\users\martin\.platformio\packages\toolchain-xtensa\xtensa-lx106-elf\include\c++\4.8.2\cwchar:44:0,
                 from c:\users\martin\.platformio\packages\toolchain-xtensa\xtensa-lx106-elf\include\c++\4.8.2\bits\postypes.h:40,
                 from c:\users\martin\.platformio\packages\toolchain-xtensa\xtensa-lx106-elf\include\c++\4.8.2\bits\char_traits.h:40,
                 from c:\users\martin\.platformio\packages\toolchain-xtensa\xtensa-lx106-elf\include\c++\4.8.2\string:40,
                 from c:\users\martin\.platformio\packages\toolchain-xtensa\xtensa-lx106-elf\include\c++\4.8.2\random:41,
                 from c:\users\martin\.platformio\packages\toolchain-xtensa\xtensa-lx106-elf\include\c++\4.8.2\bits\stl_algo.h:65,
                 from c:\users\martin\.platformio\packages\toolchain-xtensa\xtensa-lx106-elf\include\c++\4.8.2\algorithm:62,
                 from C:\Users\Martin\.platformio\packages\framework-arduinoespressif8266\cores\esp8266/Arduino.h:240,
                 from lib\NTP\src/NTP.h:6,
                 from lib\NTP\src\NTP.cpp:2:
C:\Users\Martin\.platformio\packages\framework-arduinoespressif8266\tools\sdk\libc\xtensa-lx106-elf\include/wchar.h:56:8: error: forward declaration of 'struct tm'
 struct tm;
        ^
lib\NTP\src\NTP.cpp: In member function 'int NTP::getYear()':
lib\NTP\src\NTP.cpp:175:20: error: invalid use of incomplete type 'struct tm'
     return timeinfo->tm_year + NTP_START_YEAR;
                    ^
In file included from c:\users\martin\.platformio\packages\toolchain-xtensa\xtensa-lx106-elf\include\c++\4.8.2\cwchar:44:0,
                 from c:\users\martin\.platformio\packages\toolchain-xtensa\xtensa-lx106-elf\include\c++\4.8.2\bits\postypes.h:40,
                 from c:\users\martin\.platformio\packages\toolchain-xtensa\xtensa-lx106-elf\include\c++\4.8.2\bits\char_traits.h:40,
                 from c:\users\martin\.platformio\packages\toolchain-xtensa\xtensa-lx106-elf\include\c++\4.8.2\string:40,
                 from c:\users\martin\.platformio\packages\toolchain-xtensa\xtensa-lx106-elf\include\c++\4.8.2\random:41,
                 from c:\users\martin\.platformio\packages\toolchain-xtensa\xtensa-lx106-elf\include\c++\4.8.2\bits\stl_algo.h:65,
                 from c:\users\martin\.platformio\packages\toolchain-xtensa\xtensa-lx106-elf\include\c++\4.8.2\algorithm:62,
                 from C:\Users\Martin\.platformio\packages\framework-arduinoespressif8266\cores\esp8266/Arduino.h:240,
                 from lib\NTP\src/NTP.h:6,
                 from lib\NTP\src\NTP.cpp:2:
C:\Users\Martin\.platformio\packages\framework-arduinoespressif8266\tools\sdk\libc\xtensa-lx106-elf\include/wchar.h:56:8: error: forward declaration of 'struct tm'
 struct tm;
        ^
lib\NTP\src\NTP.cpp: In member function 'int NTP::getDST()':
lib\NTP\src\NTP.cpp:199:37: error: 'ctime' was not declared in this scope
         Serial.print(ctime(&dstStart));
                                     ^
lib\NTP\src\NTP.cpp: In member function 'time_t NTP::calcTime(dstRule*)':
lib\NTP\src\NTP.cpp:240:15: error: aggregate 'tm tm2' has incomplete type and cannot be defined
     struct tm tm2;
               ^
lib\NTP\src\NTP.cpp:261:20: error: 'mktime' was not declared in this scope
     t = mktime(&tm2);        
                    ^
*** [.pio\build\d1_mini\lib9c0\NTP\NTP.cpp.o] Error 1
=========================================================================== [FAILED] Took 6.49 seconds ===========================================================================
The terminal process "C:\Users\Martin\.platformio\penv\Scripts\platformio.exe 'run'" terminated with exit code: 1.

Terminal will be reused by tasks, press any key to close it.

**Comment by micooke** [comment]
@mr-sneezy - thats an error generated by the pio core (looks like its missing #include <ctime>), I think you might need to update your pio platforms etc.

I downloaded VS Code and pulled the Platform IO plugin fresh and it compiled successfully.

```
Processing d1_mini (platform: espressif8266; board: d1_mini; framework: arduino)
----------------------------------------------------------------------------------------------------------------------------Verbose mode can be enabled via `-v, --verbose` option
CONFIGURATION: https://docs.platformio.org/page/boards/espressif8266/d1_mini.html
PLATFORM: Espressif 8266 (2.6.2) > WeMos D1 R2 and mini
HARDWARE: ESP8266 80MHz, 80KB RAM, 4MB Flash
PACKAGES: 
 - framework-arduinoespressif8266 3.20704.0 (2.7.4) 
 - tool-esptool 1.413.0 (4.13) 
 - tool-esptoolpy 1.20800.0 (2.8.0)
 - toolchain-xtensa 2.40802.200502 (4.8.2)
LDF: Library Dependency Finder -> http://bit.ly/configure-pio-ldf
LDF Modes: Finder ~ chain, Compatibility ~ soft
Found 34 compatible libraries
Scanning dependencies...
```

**Comment by mr-sneezy** [comment]
Thanks micooke. Given the news you built successfully, I setup VS Code and PlatformIO on a Win10 laptop that had never had either on it before. I cloned the repo locally and built from that successfully, and uploaded to a D1 Mini as well. So a win there.

I then tried to refresh PlatformIO on the main Win10 PC, no change. Then I uninstalled both platformIO and VS Code, restarted the PC, reinstalled both, pulled the repo down locally to a new location. Did the build, and nope failed again, with same errors. Stumped by that, but at least I can build on the laptop...
